src/blockrelay/graphene.cpp

```diff
@@ -82,6 +82,126 @@ void CGrapheneBlock::FillShortTxIDSelector()
     shorttxidk1 = shorttxidhash.GetUint64(1);
 }
 
+void CGrapheneBlock::AddNewTransactions(std::vector<CTransaction> vMissingTx, CNode *pfrom)
+{
+    if (vMissingTx.size() == 0)
+        return;
+
+    // If canonical ordering is activated, locate empty indexes in vTxHashes256 to be used in sorting
+    std::vector<size_t> missingTxIdxs;
+    if (fCanonicalTxsOrder && NegotiateGrapheneVersion(pfrom) >= 1)
+    {
+        uint256 nullhash;
+        for (size_t idx = 0; idx < vTxHashes256.size(); idx++)
+        {
+            if (vTxHashes256[idx] == nullhash)
+                missingTxIdxs.push_back(idx);
+        }
+    }
+
+    size_t idx = 0;
+    for (const CTransaction &tx : vMissingTx)
+    {
+        mapMissingTx[GetShortID(pfrom->gr_shorttxidk0, pfrom->gr_shorttxidk1, tx.GetHash(),
+            NegotiateGrapheneVersion(pfrom))] = MakeTransactionRef(tx);
+
+        uint256 hash = tx.GetHash();
+        uint64_t cheapHash =
+            GetShortID(pfrom->gr_shorttxidk0, pfrom->gr_shorttxidk1, hash, NegotiateGrapheneVersion(pfrom));
+
+        // Insert in arbitrary order if canonical ordering is enabled and xversion is recent enough
+        if (fCanonicalTxsOrder && NegotiateGrapheneVersion(pfrom) >= 1)
+            vTxHashes256[missingTxIdxs[idx++]] = hash;
+        // Otherwise, use ordering information
+        else
+            vTxHashes256[mapHashOrderIndex[cheapHash]] = hash;
+    }
+}
+
+void CGrapheneBlock::OrderTxHashes(CNode *pfrom)
+{
+    if (vTxHashes256.size() != nBlockTxs)
+        throw std::runtime_error("Cannot OrderTxHashes if size of vTxHashes256 unequal to nBlockTxs");
+
+    // Sort order transactions if canonical order is enabled and graphene version is late enough
+    if (fCanonicalTxsOrder && NegotiateGrapheneVersion(pfrom) >= 1)
+    {
+        // coinbase is always first
+        std::sort(vTxHashes256.begin() + 1, vTxHashes256.end());
+        LOG(GRAPHENE, "Using canonical order for block from peer=%s\n", pfrom->GetLogName());
+    }
+    else
+    {
+        uint256 nullhash;
+        std::vector<uint256> orderedTxHashes256(nBlockTxs, nullhash);
+        for (auto &hash : vTxHashes256)
+        {
+            uint64_t cheapHash =
+                GetShortID(pfrom->gr_shorttxidk0, pfrom->gr_shorttxidk1, hash, NegotiateGrapheneVersion(pfrom));
+            orderedTxHashes256[mapHashOrderIndex[cheapHash]] = hash;
```

Comment:
here as well, I'm not convinced that mapHashOrderIndex[cheapHash] is never null, although I haven't chased around to discover that.

Reply:
I replaced this line with a mapHashOrderIndex.find, I throw if not found, otherwise index on what's found.